### DISCUSSION TITLE
`brew install libssh2` error  'libssh2: Failed to download resource "libssh2.rb"'

### DISCUSSION THREAD

**Discussion by uliyt** [Homebrew/discussions#5724] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 824efa8836dc226aa92dbbf7404c1b4a66707cca
Last commit: 4 days ago
Core tap JSON: 08 Nov 07:26 UTC
Core cask tap JSON: 08 Nov 07:26 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 3.3.5 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 12.7.2-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libssh2.1.0.1.dylib
  /usr/local/lib/libssh2.1.dylib
  /usr/local/lib/libtcl8.6.dylib
  /usr/local/lib/libtk8.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected header files:
  /usr/local/include/fakemysql.h
  /usr/local/include/fakepq.h
  /usr/local/include/fakesql.h
  /usr/local/include/itcl.h
  /usr/local/include/itcl2TclOO.h
  /usr/local/include/itclDecls.h
  /usr/local/include/itclInt.h
  /usr/local/include/itclIntDecls.h
  /usr/local/include/itclMigrate2TclCore.h
  /usr/local/include/itclTclIntStubsFcn.h
  /usr/local/include/libssh2.h
  /usr/local/include/libssh2_publickey.h
  /usr/local/include/libssh2_sftp.h
  /usr/local/include/mysqlStubs.h
  /usr/local/include/odbcStubs.h
  /usr/local/include/pqStubs.h
  /usr/local/include/tcl.h
  /usr/local/include/tclDecls.h
  /usr/local/include/tclOO.h
  /usr/local/include/tclOODecls.h
  /usr/local/include/tclPlatDecls.h
  /usr/local/include/tclThread.h
  /usr/local/include/tclTomMath.h
  /usr/local/include/tclTomMathDecls.h
  /usr/local/include/tdbc.h
  /usr/local/include/tdbcDecls.h
  /usr/local/include/tdbcInt.h
  /usr/local/include/tk.h
  /usr/local/include/tkDecls.h
  /usr/local/include/tkPlatDecls.h

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libssh2.la

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/libssh2.pc
  /usr/local/lib/pkgconfig/tcl.pc
  /usr/local/lib/pkgconfig/tk.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libssh2.a
  /usr/local/lib/libtclstub8.6.a
  /usr/local/lib/libtkstub8.6.a
```

### Description of issue

I am installing libssh2.

First, `brew install libssh2` returned error message: 
`Error: libssh2: Failed to download resource "libssh2.rb"
Download failed: https://raw.githubusercontent.com/Homebrew/homebrew-core/0d27f082f92021422f36a291d3795c8e59ea7eda/Formula/lib/libssh2.rb`

Then, I configured libssh2 locally, the files are configured to `/usr/local/lib/`. However, `brew info libssh2`  returned message:
`==> libssh2: stable 1.11.1, HEAD
C library implementing the SSH2 protocol
https://libssh2.org/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/lib/libssh2.rb
License: BSD-3-Clause
==> Dependencies
Required: openssl@3 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 89,366 (30 days), 163,824 (90 days), 551,599 (365 days)
install-on-request: 11,827 (30 days), 14,175 (90 days), 59,860 (365 days)
build-error: 31 (30 days)`

Lastly, I tried to fetch libssh2 first and then install it:
`brew fetch --debug --verbose libssh2`, this command successfully downloaded `libssh2.tar.gz` file.
`brew install libssh2` still failed and returned the same error message.

Appreciate any suggestions.

**Comment by SMillerDev** [comment]
Can you give the full output of the command?

**Comment by uliyt** [comment]
Output from command `$ brew fetch --debug --verbose libssh2`:
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromAPILoader): loading libssh2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::NullLoader): loading libssh2
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.4.4\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 12.7.2\)\ curl/8.1.2 --header Accept-Language:\ en --retry 3 -V
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.4.4\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 12.7.2\)\ curl/8.1.2 --header Accept-Language:\ en --retry 3 --fail --location --silent --head https://libssh2.org/download/libssh2-1.11.1.tar.gz
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.4.4\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 12.7.2\)\ curl/8.1.2 --header Accept-Language:\ en --retry 3 --fail --location --silent --head --request GET https://libssh2.org/download/libssh2-1.11.1.tar.gz
==> Downloading https://libssh2.org/download/libssh2-1.11.1.tar.gz
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.4.4\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 12.7.2\)\ curl/8.1.2 --header Accept-Language:\ en --fail --retry 3 --remote-time --output /Users/user0/Library/Caches/Homebrew/downloads/70868fcb9963c288f3b0f9e961ebc21b984fcd4ebc1c9399e5b6a0ecd68653c4--libssh2-1.11.1.tar.gz.incomplete --location https://libssh2.org/download/libssh2-1.11.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1067k  100 1067k    0     0  13206      0  0:01:22  0:01:22 --:--:-- 86994
Downloaded to: /Users/user0/Library/Caches/Homebrew/downloads/70868fcb9963c288f3b0f9e961ebc21b984fcd4ebc1c9399e5b6a0ecd68653c4--libssh2-1.11.1.tar.gz
SHA256: d9ec76cbe34db98eec3539fe2c899d26b0c837cb3eb466a56b0f109cabf658f7
==> Verifying checksum for '70868fcb9963c288f3b0f9e961ebc21b984fcd4ebc1c9399e5b6a0ecd68653c4--libssh2-1.11.1.tar.gz'



Output from command `$ brew install libssh2`:
==> Downloading https://formulae.brew.sh/api/formula.jws.json
######################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
######################################################################### 100.0%
==> Fetching dependencies for libssh2: ca-certificates
==> Fetching ca-certificates
==> Downloading https://ghcr.io/v2/homebrew/core/ca-certificates/manifests/2024-
Already downloaded: /Users/user0/Library/Caches/Homebrew/downloads/338dad7c2ff7c822cda7c417944521589856741c0fbd7a7f07b88a18d7fb7e05--ca-certificates-2024-09-24.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/ca-certificates/blobs/sha256:21
Already downloaded: /Users/user0/Library/Caches/Homebrew/downloads/a42908db56776e38039ac5f25968df8df16307650c519973b7d5ed689a293c88--ca-certificates--2024-09-24.all.bottle.tar.gz
==> Fetching libssh2
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/04269cf
curl: (7) Failed to connect to raw.githubusercontent.com port 443 after 21 ms: Couldn't connect to server

Error: libssh2: Failed to download resource "libssh2.rb"
Download failed: https://raw.githubusercontent.com/Homebrew/homebrew-core/04269cffc62f2e5cbdebf1a20a4e97b7c4e71132/Formula/lib/libssh2.rb

**Reply by SMillerDev** [reply]
Like I suspected, there is a problem with your connection to GitHub. That's unfortunately something we can't really help with. It's between your machine and GitHub.

**Reply by uliyt** [reply]
That reminds something that I have done recently to my git. Will confirm that and leave a solution here if it is a common issue.

**Reply by uliyt** [reply]
Well the problem is still there. My git now can clone with url starts with https (previously can only clone with ssh), but the brew install still fails.

**Reply by ZhongRuoyu** [reply]
You need to diagnose your connection to raw.githubusercontent.com (i.e., make sure `curl https://raw.githubusercontent.com/Homebrew/homebrew-core/04269cffc62f2e5cbdebf1a20a4e97b7c4e71132/Formula/lib/libssh2.rb` works). It's possible that you can connect to github.com but not raw.githubusercontent.com.

**Reply by uliyt** [reply]
Yes, `curl` works but not between `curl` and `https://raw.githubusercontent.com/Homebrew/homebrew-core/04269cffc62f2e5cbdebf1a20a4e97b7c4e71132/Formula/lib/libssh2.rb`.

**Reply by uliyt** [reply]
Output of `curl -O https://raw.githubusercontent.com/Homebrew/homebrew-core/04269cffc62f2e5cbdebf1a20a4e97b7c4e71132/Formula/lib/libssh2.rb`:

% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to raw.githubusercontent.com port 443 after 21 ms: Couldn't connect to server

**Reply by gromgit** [reply]
Works for me, and since `curl` failed after just 21ms, it suggests there's a firewall or some other network device between you and that site that's actively dropping your connections.

One other thing to check: What's the output of `dig raw.githubusercontent.com`?

**Reply by uliyt** [reply]
Output of `dig raw.githubusercontent.com`:

$ dig raw.githubusercontent.com

; <<>> DiG 9.10.6 <<>> raw.githubusercontent.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45223
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;raw.githubusercontent.com.	IN	A

;; ANSWER SECTION:
raw.githubusercontent.com. 300	IN	A	0.0.0.0

;; Query time: 30 msec
;; SERVER: 192.168.1.1#53(192.168.1.1)
;; WHEN: Tue Nov 12
;; MSG SIZE  rcvd: 59

**Comment by uliyt** [comment]
Add the correct `ip` of `raw.githubusercontent.com` to hosts file will solve this.

My added line is: `199.232.68.133 raw.githubusercontent.com`

Will close this thread.

**Reply by gromgit** [reply]
Instead of adding a fixed IP mapping, which may change at any time, why not _remove_ that entry from the DNS server on 192.168.1.1 and let it resolve normally?

**Reply by uliyt** [reply]
There are two entries of DNS server on this machine, they are not removable. I have attached a image below.

![image](https://github.com/user-attachments/assets/164afde8-6360-4f65-a3c8-adfafd812a91)

**Reply by gromgit** [reply]
I'm referring to the DNS server at 192.168.1.1 (likely your Internet gateway), which is telling you that `raw.githubusercontent.com` is at IP `0.0.0.0`. That's wrong in every possible way, and should be corrected _at the gateway_, rather than having to override it on your own machine.
